app/src/components/features/rules/RuleBuilder/Header/ActionButtons/ShareRuleButton/index.jsx

```diff
@@ -1,30 +1,103 @@
-import React from "react";
+import React, { useState } from "react";
+import { useSelector, useDispatch } from "react-redux";
+import { actions } from "store";
+import { getCurrentlySelectedRuleData, getUserAuthDetails } from "store/selectors";
 import { RQButton } from "lib/design-system/components";
-import { Tooltip } from "antd";
+import { Button, Tooltip } from "antd";
+import CreateSharedListModal from "components/features/sharedLists/CreateSharedListModal";
+import { fetchSharedLists } from "components/features/sharedLists/SharedListsIndexPage/actions";
 import { trackRuleEditorHeaderClicked } from "modules/analytics/events/common/rules";
+import { AUTH } from "modules/analytics/events/common/constants";
+import APP_CONSTANTS from "config/constants";
 import { getModeData } from "../../../actions";
 
-const ShareRuleButton = ({ handleShareRuleClick, ruleType }) => {
+const ShareRuleButton = ({ isRuleEditorModal }) => {
   const { MODE } = getModeData(window.location);
+  const [isShareRulesModalActive, setIsShareRulesModalActive] = useState(false);
+  const user = useSelector(getUserAuthDetails);
+  const dispatch = useDispatch();
+  const currentlySelectedRuleData = useSelector(getCurrentlySelectedRuleData);
+
+  const toggleShareRulesModal = () => {
+    setIsShareRulesModalActive((prev) => !prev);
+  };
+
+  const shareRuleClickHandler = () => {
+    if (user.loggedIn) {
+      verifySharedListsLimit();
+    } else {
+      dispatch(
+        actions.toggleActiveModal({
+          modalName: "authModal",
+          newValue: true,
+          authMode: APP_CONSTANTS.AUTH.ACTION_LABELS.SIGN_UP,
+          eventSource: AUTH.SOURCE.SHARE_RULES,
+        })
+      );
+    }
+  };
+
+  const verifySharedListsLimit = () => {
+    dispatch(
+      actions.toggleActiveModal({
+        modalName: "loadingModal",
+        newValue: true,
+      })
+    );
+
+    fetchSharedLists(user.details.profile.uid).then((result) => {
+      //Create new shared list
+      toggleShareRulesModal();
+      //Deactivate loading modal
+      dispatch(
+        actions.toggleActiveModal({
+          modalName: "loadingModal",
+          newValue: false,
+        })
+      );
+    });
+  };
   return (
-    <Tooltip title="Share rule" placement="bottom">
-      <RQButton
-        type="primary"
-        icon={
-          <img
-            width="13.4px"
-            height="10px"
-            alt="down arrow"
-            src="/assets/icons/share.svg"
-            className="rule-header-share-btn-icon"
+    <>
+      {isRuleEditorModal ? (
+        <Button
+          type="text"
+          onClick={() => {
+            shareRuleClickHandler();
+            trackRuleEditorHeaderClicked("share_button", currentlySelectedRuleData.ruleType, MODE);
```

Comment:
please add `source` property in this and rest of the actions `value: modal | screen`